evalml/tests/component_tests/test_imputer.py

```diff
@@ -339,39 +356,137 @@ def test_imputer_no_nans(imputer_test_data):
-    assert_frame_equal(expected, transformed, check_dtype=False)
+expected = pd.DataFrame(
+    {
+        "int with None": pd.Series([1, 0, 5, 5, 5], dtype="Int64"),
+        "float with None": [0.1, 0.0, 0.5, 0.5, 0.5],
+        "category with None": pd.Series(["c", "a", "a", "a", "a"], dtype="category"),
+        "boolean with None": pd.Series(
+            [True, True, False, True, True], dtype="boolean"
+        ),
+        "object with None": pd.Series(["b", "a", "a", "a", "a"], dtype="category"),
+    }
+)
+ltypes = {
+    "int with None": "IntegerNullable",
+    "float with None": "Double",
+    "category with None": "categorical",
+    "boolean with None": "BooleanNullable",
+    "object with None": "categorical",
+    "all None": "categorical",
+}
+columns_dict = {
+    "integers_only": ["int with None"],
+    "floats_only": ["float with None"],
+    "numerics_only": ["int with None", "float with None"],
+    "categoricals_only": ["category with None", "object with None"],
+    "booleans_only": ["boolean with None"],
+    "categorical_and_booleans": [
+        "category with None",
+        "boolean with None",
+        "object with None",
+    ],
+    "all": [
+        "int with None",
+        "float with None",
+        "category with None",
+        "boolean with None",
+        "object with None",
+    ],
+}
+
+
+@pytest.mark.parametrize(
+    "dtypes",
+    [
+        "integers_only",
+        "floats_only",
+        "numerics_only",
+        "booleans_only",
+        "categoricals_only",
+        "categorical_and_booleans",
+        "all",
+    ],
+)
+@pytest.mark.parametrize(
+    "numeric_impute_strategy", ["most_frequent", "mean", "median", "constant"]
+)
+@pytest.mark.parametrize("categorical_impute_strategy", ["most_frequent", "constant"])
+@pytest.mark.parametrize("boolean_impute_strategy", ["most_frequent", "constant"])
+def test_imputer_with_none_separated(
+    dtypes,
+    numeric_impute_strategy,
+    categorical_impute_strategy,
+    boolean_impute_strategy,
+):
+    """Test for correctness for the different numeric, categorical and boolean dtypes using dataframes that contain
+    either just the tested imputed dtypes or combinations of dtypes."""
+
+    test_ltypes = dict((k, ltypes[k]) for k in columns_dict[dtypes])
+    X_test = X[columns_dict[dtypes]]
+    X_test.ww.init(logical_types=test_ltypes)
+    y = pd.Series([0, 0, 1, 0, 1])
+    numeric_fill_value = 0
+    categorical_fill_value = "filler"
+    boolean_fill_value = True
+    imputer = Imputer(
+        numeric_impute_strategy=numeric_impute_strategy,
+        categorical_impute_strategy=categorical_impute_strategy,
+        numeric_fill_value=numeric_fill_value,
+        categorical_fill_value=categorical_fill_value,
+        boolean_impute_strategy=boolean_impute_strategy,
+        boolean_fill_value=boolean_fill_value,
+    )
+    imputer.fit(X_test, y)
+    transformed = imputer.transform(X_test, y)
+
+    # Build the expected dataframe
```

Comment:
I don't really like the way I'm doing this here - it feels clunky, but I've been looking at it too long.  Basically I'm operating under the assumption that the last value is the value to impute and I'm changing out the last row of the dataframe with the expected imputed results for each test.  Additionally, I don't like that the expected data frame, without modifications, is the answer for the default set of tests.  Open to ideas!

Reply:
In my opinion, I think the `except KeyError` makes it harder to understand what's happening. I think `for col in column_dict[...] if col in expected_columns` makes the intention more explicit

Reply:
But other than that, I think it's the classic trade-off between coverage and complexity. I think having coverage for all the combinations of parameters is valuable and unless we pre-specify the answer for all possible combinations, we'll have to build it on the spot. Along those lines though, maybe we can just pre-compute the expected last row for each combination and then just replace that row in the copy of `expected`

Reply:
Thanks for the quality improvement!